File: src/Vograph.Core/Services/TimetableApiCache.cs
using Vograph.Core.Models;
using System.Text.Json;
using Microsoft.Data.Sqlite;

namespace Vograph.Core.Services;

public sealed class TimetableApiCache(Database db)
{
    public static void EnsureSchema(SqliteConnection connection)
    {
        using var cmd = connection.CreateCommand();
        cmd.CommandText = """
            CREATE TABLE IF NOT EXISTS api_catalog (groupId TEXT PRIMARY KEY, name TEXT NOT NULL);
            CREATE TABLE IF NOT EXISTS api_cache_metadata (groupId TEXT PRIMARY KEY, snapshotId TEXT, payload TEXT NOT NULL);
            """;
        cmd.ExecuteNonQuery();
    }

    public CacheMetadata? Read(string? groupId)
    {
        if (groupId is null) return null;
        using var cmd = db.Connection.CreateCommand();
        cmd.CommandText = "SELECT payload FROM api_cache_metadata WHERE groupId=@id";
        cmd.Parameters.AddWithValue("@id", groupId);
        return cmd.ExecuteScalar() is string json ? JsonSerializer.Deserialize<CacheMetadata>(json) : null;
    }

    public bool CanIntersect(string selected, string friend)
    {
        var mine = Read(selected);
        var other = Read(friend);
        if (mine?.Source != "api" && other?.Source != "api") return true;
        return mine?.Source == "api" && other?.Source == "api" && mine.SourceBase == other.SourceBase &&
            mine.Period == other.Period && mine.Meta?.SnapshotId == other.Meta?.SnapshotId &&
            mine.Meta?.Stale == false && other.Meta?.Stale == false;
    }

    public void Apply(TimetableApiSnapshot snapshot, string sourceBase = "")
    {
        Validate(snapshot);
        using var tx = db.Connection.BeginTransaction();
        // Capture old legacy periods before any later settings save can copy the selected API period.
        var settings = db.GetSettings();
        var firstAdoption = Read("") is null;
        var oldPeriod = new TimetableApiPeriod(DateOnly.TryParse(settings.PeriodStart, out var start) ? start : new(DateTime.Now.Year, 9, 1),
            settings.WeekCount, settings.PeriodTitle ?? "", "Europe/Moscow");
        using (var cmd = db.Connection.CreateCommand())
        {
            cmd.Transaction = tx;
            cmd.CommandText = "SELECT id, lastFetchedAt FROM groups";
            var old = new List<(string Id, string? Fetched)>();
            using (var reader = cmd.ExecuteReader())
                while (reader.Read()) old.Add((reader.GetString(0), reader.IsDBNull(1) ? (firstAdoption ? settings.LastFetchedAt : null) : reader.GetString(1)));
            foreach (var group in old)
                if (Read(group.Id) is null && (group.Fetched is not null || db.GetAllLessonsForGroup(group.Id).Count > 0))
                    Write(group.Id, new(oldPeriod, null, group.Fetched, "legacy", ""), tx);
        }
        Execute("DELETE FROM api_catalog", tx);
        foreach (var group in snapshot.Groups)
        {
            Execute("INSERT INTO groups(id,name) VALUES(@id,@name) ON CONFLICT(id) DO UPDATE SET name=excluded.name", tx,
                ("@id", group.Id), ("@name", group.Name));
            Execute("INSERT INTO api_catalog(groupId,name) VALUES(@id,@name)", tx, ("@id", group.Id), ("@name", group.Name));
        }
        Write("", new(snapshot.Period, snapshot.Meta, snapshot.Meta.FetchedAt.ToString("o"), "api", sourceBase), tx);
        foreach (var (id, downloaded) in snapshot.DownloadedGroups.OrderBy(x => x.Key, StringComparer.Ordinal))
        {
            var group = downloaded.Group.ToGroup();
            Execute("UPDATE groups SET rawXml=NULL,url='',lastFetchedAt=@at WHERE id=@id", tx,
                ("@at", downloaded.Meta.FetchedAt.ToString("o")), ("@id", group.Id));
            db.ClearScheduleForGroup(id);
            foreach (var lesson in downloaded.Lessons) db.InsertLesson(lesson.ToLesson(id));
            Write(id, new(snapshot.Period, downloaded.Meta, downloaded.Meta.FetchedAt.ToString("o"), "api", sourceBase), tx);
        }
        Execute("UPDATE settings SET lastAutoCheckAt=@at WHERE id=1", tx, ("@at", DateTime.UtcNow.ToString("o")));
        tx.Commit();
        // Derived homework is deliberately outside this transaction; caller logs recompute failures safely.
    }

    private void Write(string id, CacheMetadata metadata, SqliteTransaction tx) => Execute(
        "INSERT INTO api_cache_metadata(groupId,snapshotId,payload) VALUES(@id,@snapshot,@payload) ON CONFLICT(groupId) DO UPDATE SET snapshotId=excluded.snapshotId,payload=excluded.payload",
        tx, ("@id", id), ("@snapshot", metadata.Meta?.SnapshotId.ToString()), ("@payload", JsonSerializer.Serialize(metadata)));

    private void Execute(string sql, SqliteTransaction tx, params (string Name, string? Value)[] parameters)
    {
        using var cmd = db.Connection.CreateCommand();
        cmd.Transaction = tx;
        cmd.CommandText = sql;
        foreach (var (name, value) in parameters) cmd.Parameters.AddWithValue(name, (object?)value ?? DBNull.Value);
        cmd.ExecuteNonQuery();
    }

    private static void Validate(TimetableApiSnapshot snapshot)
    {
        // Trusted transport boundary, not a second JSON validator. Public records must not poison keys/generation.
        if (snapshot.Meta.SnapshotId == Guid.Empty || snapshot.Groups.IsDefault || snapshot.Period.WeekCount != 2 ||
            snapshot.Period.TimeZone != "Europe/Moscow" || snapshot.Groups.Any(g => string.IsNullOrWhiteSpace(g.Id) || g.LessonCount < 0) ||
            snapshot.Groups.Select(g => g.Id).Distinct(StringComparer.Ordinal).Count() != snapshot.Groups.Length)
            throw new TimetableApiException(TimetableApiFailure.InvalidPayload);
        foreach (var (id, group) in snapshot.DownloadedGroups)
            if (group.Group.Id != id || !snapshot.Groups.Contains(group.Group) || group.Meta.SnapshotId != snapshot.Meta.SnapshotId ||
                group.Lessons.IsDefault || group.Lessons.Length != group.Group.LessonCount)
                throw new TimetableApiException(TimetableApiFailure.InvalidPayload);
    }
}

public sealed record CacheMetadata(TimetableApiPeriod Period, TimetableApiMeta? Meta, string? FetchedAt, string Source, string SourceBase);

File: android/app/src/main/java/ru/bgtu_voenmeh/zapara/data/api/TimetableApiCache.kt
package ru.bgtu_voenmeh.zapara.data.api

import ru.bgtu_voenmeh.zapara.data.GroupInfo
import java.time.LocalDate

class TimetableApiCache(private val store: TimetableStore) {
    fun read(groupId: String?): CacheMetadata? = groupId?.let { store.readMetadata(it) }

    fun canIntersect(selected: String, friend: String): Boolean {
        val mine = read(selected)
        val other = read(friend)
        if (mine?.source != "api" && other?.source != "api") return true
        return mine?.source == "api" && other?.source == "api" && mine.sourceBase == other.sourceBase &&
            mine.period == other.period && mine.meta?.snapshotId == other.meta?.snapshotId &&
            mine.meta?.stale == false && other.meta?.stale == false
    }

    fun apply(snapshot: TimetableApiSnapshot, sourceBase: String = "") {
        validate(snapshot)
        store.runInTransaction {
            val settings = store.settings()
            val firstAdoption = read("") == null
            val oldPeriod = TimetableApiPeriod(
                settings.periodStart,
                settings.weekCount,
                settings.periodTitle.orEmpty(),
                "Europe/Moscow"
            )
            for (group in store.storedGroups()) {
                val fetched = if (firstAdoption) settings.lastFetchedAt else null
                if (read(group.id) == null && (fetched != null || store.allLessons(group.id).isNotEmpty())) {
                    store.writeMetadata(group.id, CacheMetadata(oldPeriod, null, fetched, "legacy", ""))
                }
            }
            store.clearCatalog()
            for (group in snapshot.groups) {
                store.ensureGroup(group.id, group.name)
                store.insertCatalog(group.id, group.name)
            }
            store.writeMetadata(
                "",
                CacheMetadata(snapshot.period, snapshot.meta, snapshot.meta.fetchedAt, "api", sourceBase)
            )
            for ((id, downloaded) in snapshot.downloaded.entries.sortedBy { it.key }) {
                store.ensureGroup(id, downloaded.group.name, "")
                store.replaceLessons(id, downloaded.lessons.map { it.toLesson(id) })
                store.writeMetadata(
                    id,
                    CacheMetadata(snapshot.period, downloaded.meta, downloaded.meta.fetchedAt, "api", sourceBase)
                )
            }
        }
    }

    private fun validate(snapshot: TimetableApiSnapshot) {
        if (snapshot.meta.snapshotId == "00000000-0000-0000-0000-000000000000" ||
            snapshot.period.weekCount != 2 ||
            snapshot.period.timeZone != "Europe/Moscow" ||
            snapshot.groups.any { it.id.isBlank() || it.lessonCount < 0 } ||
            snapshot.groups.map { it.id }.distinct().size != snapshot.groups.size
        ) {
            throw TimetableApiException(TimetableApiFailure.InvalidPayload)
        }
        for ((id, group) in snapshot.downloaded) {
            if (group.group.id != id || group.group !in snapshot.groups ||
                group.meta.snapshotId != snapshot.meta.snapshotId ||
                group.lessons.size != group.group.lessonCount
            ) {
                throw TimetableApiException(TimetableApiFailure.InvalidPayload)
            }
        }
    }
}

fun defaultPeriodStart(): LocalDate = LocalDate.of(2026, 9, 1)

internal fun CacheMetadata.encode(): String = buildString {
    append('{')
    append("\"period\":{\"start\":\"").append(period.start).append("\",\"weekCount\":").append(period.weekCount)
    append(",\"title\":").append(jsonEsc(period.title)).append(",\"timeZone\":").append(jsonEsc(period.timeZone)).append("},")
    append("\"meta\":")
    if (meta == null) append("null") else {
        append("{\"snapshotId\":").append(jsonEsc(meta.snapshotId))
        append(",\"fetchedAt\":").append(jsonEsc(meta.fetchedAt))
        append(",\"publishedAt\":").append(jsonEsc(meta.publishedAt))
        append(",\"sourceModifiedAt\":")
        if (meta.sourceModifiedAt == null) append("null") else append(jsonEsc(meta.sourceModifiedAt))
        append(",\"sourceKind\":").append(jsonEsc(meta.sourceKind))
        append(",\"sourceUrl\":")
        if (meta.sourceUrl == null) append("null") else append(jsonEsc(meta.sourceUrl))
        append(",\"sourceSha256\":").append(jsonEsc(meta.sourceSha256))
        append(",\"stale\":").append(meta.stale).append('}')
    }
    append(",\"fetchedAt\":")
    if (fetchedAt == null) append("null") else append(jsonEsc(fetchedAt))
    append(",\"source\":").append(jsonEsc(source))
    append(",\"sourceBase\":").append(jsonEsc(sourceBase))
    append('}')
}

private fun jsonEsc(value: String): String = buildString {
    append('"')
    for (c in value) when (c) {
        '\\' -> append("\\\\")
        '"' -> append("\\\"")
        else -> append(c)
    }
    append('"')
}

internal fun decodeCacheMetadata(payload: String): CacheMetadata {
    val root = StrictJson.parse(payload).obj()
    val periodObj = root.field("period").obj()
    val period = TimetableApiPeriod(
        LocalDate.parse(periodObj.text("start", 10)),
        periodObj.int("weekCount"),
        periodObj.text("title", 256),
        periodObj.text("timeZone", 64)
    )
    val meta = if (root.field("meta") is JsonValue.Null) null else {
        val m = root.field("meta").obj()
        TimetableApiMeta(
            m.text("snapshotId", 36),
            m.text("fetchedAt", 40),
            m.text("publishedAt", 40),
            m.nullableText("sourceModifiedAt", 40),
            m.text("sourceKind", 64),
            m.nullableText("sourceUrl"),
            m.text("sourceSha256", 64),
            m.bool("stale")
        )
    }
    return CacheMetadata(period, meta, root.nullableText("fetchedAt", 40), root.text("source", 32), root.text("sourceBase", 2048))
}
